include/mbedtls/fsio.h

```diff
@@ -0,0 +1,252 @@
+/**
+ * \file fsio.h
+ *
+ * \brief File read/write functions
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+#ifndef MBEDTLS_FILE_H
+#define MBEDTLS_FILE_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+/* If MBEDTLS_FS_IO is enabled then file IO functions should be made available
+ * via standard library or platform specific implementation. */
+
+/* Default and alternative implementation specific interfaces. */
+#if !defined(MBEDTLS_FS_IO_ALT)
+#include <stdio.h>
+#include <dirent.h>
+
+typedef FILE *  mbedtls_file_t;
+#define mbedtls_fread       fread
+#define mbedtls_fgets       fgets
+#define mbedtls_fwrite      fwrite
+#define mbedtls_fclose      fclose
+#define mbedtls_ferror      ferror
+#define mbedtls_fseek       fseek
+#define mbedtls_ftell       ftell
+#define MBEDTLS_SEEK_SET    SEEK_SET
+#define MBEDTLS_SEEK_CUR    SEEK_CUR
+#define MBEDTLS_SEEK_END    SEEK_END
+#define MBEDTLS_FILE_INVALID    NULL
+
+typedef DIR *   mbedtls_dir_t;
+#define mbedtls_opendir         opendir
+#define mbedtls_closedir        closedir
+#define MBEDTLS_FSIO_DT_BLK     DT_BLK
```

Comment:
Why expose so many different file types? We aren't trying to provide a POSIX abstraction layer here. Any filesystem element should be reported as a directory if it's a directory (or a symlink to one) or as a file if it's any kind of file that you can read and write data from. Possibly add a third type “unknown” for things that have a name but don't fit as files or directories, such as named sockets, doors and broken symlinks.

Reply:
Do you mean to say apart from ```DT_REG``` and possibly ```DT_DIR``` we should say ```DT_UNKNOWN```? Shall we call ```DT_UNKNOWN``` as ```DT_UNSUPPORTED```?

Reply:
@mazimkhan I think there should be the following three possibilities: what you can fopen, what you can opendir, and what you can't use at all through the mbedtls abstraction.

    enum {
        MBEDTLS_FSIO_DT_FILE, // A file can be opened with `mbedtls_fopen` and then read and written as a stream of bytes. Some files may unreadable or unwritable due to their nature (e.g. unidirectional pipe), permissions or other system restrictions. Files are not necessarily seekable.
        MBEDTLS_FSIO_DT_DIRECTORY, // A directory can be opened with `mbedtls_opendir` and then listed.
        MBEDTLS_FSIO_DT_OTHER, // There is a directory by this name, but it does not correspond to any type of directory entry that the library support.
    }

Names and presentation up to you, but I wouldn't call the “file” entry “regular”, because it includes things like pipes.